### PR TITLE
Ion storms (mostly) blockable by Spatial Interdictor

### PR DESCRIPTION
**Pull request by Kubius**
[OBJECTS][EVENTS][BALANCE]
## About the PR 

Fixes ion storms being absent from spatial interdictors' gamut of protection.

When an ion category is picking target objects, it will check for interdictors in range of the object before applying the malfunction behavior, and try to spend an interdiction cost based on the ion category (using a new interdict_cost variable in it, defaulting to 250).

Successful interdiction will prevent the malfunction behavior from kicking in, and make a small sparking noise from the ionization being counteracted.

**Law racks are deliberately not protected.** In addition to being sensitive electronics that are much harder to shield, there's a gameplay reason for this; allowing the interdictor to prevent ion storms could make it a default course of action to put an interdictor by the AI law rack, making it difficult for the AI to ever actually get to play around with the scenarios the alternate laws provide.

## Why's this needed? 

Assorted system malfunctions caused by ion storms should likely be within the range of things an interdictor can protect from.

## Changelog 

```changelog
(u)Kubius
(+)Spatial interdictors can now protect most devices from ion storms if the affected device is within interdiction range. Law racks are an exception, due to the sensitive nature of involved components.
```

### CODE REVIEW

File: code/modules/events/ion_storm.dm [hunk 1]
```diff
@@ -178,13 +179,29 @@ ABSTRACT_TYPE(/datum/ion_category)
 		if (!length(targets))
 			build_targets()
 		for (var/i in 1 to amount)
-			var/object = pick(targets)
+			var/atom/object = pick(targets)
+
+			//spatial interdictor: shield general hardware from ionic interference. law racks explicitly omitted due to sensitivity (and gameplay fun)
+			//consumes cell charge per hardware item protected, based on the category's interdict cost
+			var/interdicted = FALSE
+			for (var/obj/machinery/interdictor/IX in by_type[/obj/machinery/interdictor])
```
Comment (pali6) [inline code/modules/events/ion_storm.dm:187]:
```suggestion
			for_by_tcl(IX, /obj/machinery/interdictor)
```

File: code/modules/events/ion_storm.dm [hunk 2]
```diff
@@ -178,13 +179,29 @@ ABSTRACT_TYPE(/datum/ion_category)
...
+				if (IN_RANGE(IX,object,IX.interdict_range) && IX.expend_interdict(interdict_cost))
+					interdicted = TRUE
+					SPAWN(rand(1,8))
+						playsound(object.loc, "sparks", 60, 1) //absorption noise, as a little bit of "force feedback"
+					//in case people say a thing was ion stormed but it actually was not because of interdiction
```
Comment (pali6) [inline code/modules/events/ion_storm.dm:192]: I don't think this is really necessary. Targets are already picked randomly so this is just equivalent to the target not getting picked.
```suggestion
```

File: code/modules/events/ion_storm.dm [hunk 3]
```diff
@@ -178,13 +179,29 @@ ABSTRACT_TYPE(/datum/ion_category)
...
-			var/object = pick(targets)
+			var/atom/object = pick(targets)
+
+			//spatial interdictor: shield general hardware from ionic interference. law racks explicitly omitted due to sensitivity (and gameplay fun)
+			//consumes cell charge per hardware item protected, based on the category's interdict cost
+			var/interdicted = FALSE
+			for (var/obj/machinery/interdictor/IX in by_type[/obj/machinery/interdictor])
+				if (IN_RANGE(IX,object,IX.interdict_range) && IX.expend_interdict(interdict_cost))
+					interdicted = TRUE
+					SPAWN(rand(1,8))
```
Comment (pali6) [inline code/modules/events/ion_storm.dm:190]: Is a spawn really necessary there? Could we just play the sound right away?

**Review by pali6 — CHANGES_REQUESTED** [review]

File: code/modules/events/ion_storm.dm (hunk 2 above)
Comment (Kubius) [reply code/modules/events/ion_storm.dm]: Removed in [c923f6c](https://github.com/goonstation/goonstation/pull/11870/commits/c923f6ce9d6d7c4835c7dd74ce7dab91928a5230).

File: code/modules/events/ion_storm.dm (hunk 3 above)
Comment (Kubius) [reply code/modules/events/ion_storm.dm:190]: Ionizations of each category happen in batches with no time gap between targets, making it so all the zaps happen at the same time; deferring them a bit gives more of the impression of the effect being deflected and dissipated.

**Pull request merged**